pkg/placement/placement-resolver.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// A PlacementResolver holds a collection of placement resolutions.
+// The collection is indexed by placementKey string, the resolver does not
+// care what the strings are, but a string should be a placement's unique
+// identifier. The resolution for a given key can be updated,
+// exported and compared to the PlacementDecision representation.
+type PlacementResolver interface {
+	// GeneratePlacementDecision returns the placement decision for the given
+	// placement key. If a key does not exist in the cache, it is created.
+	// This function can fail if any object's GVR fails to be retrieved given
+	// its GVK, by the mapper shared with the placement-controller.
+	GeneratePlacementDecision(placementKey string) (*v1alpha1.PlacementDecisionSpec, error)
+	// ComparePlacementDecision compares the given placement decision spec
+	// with the maintained placement decision for the given placement key.
+	// The returned value is true iff:
+	//
+	// - The destinations in the PlacementDecisionSpec are an exact match
+	//of those in the resolution.
+	//
+	// - The same is true for every selected object.
+	//
+	// It is possible to output a false negative due to a temporary state of
+	// internal mappings being out of sync.
+	ComparePlacementDecision(placementKey string,
+		placementDecisionSpec *v1alpha1.PlacementDecisionSpec) bool
+	// NoteObject updates the maintained placement's objects resolution for the
+	// given placement key. If the object is being deleted, it is removed from
+	// the resolution if exists.
+	//
+	// If a key does not exist, it is created.
```

Comment:
(A) Since the caller is providing the key, it certainly already exists. I suppose this comment is about associating a resolution with the given key.

(B) What is the value in creating an empty resolution and associating it with the key in this case? Why not just return `false`?